### ISSUE TITLE
Case-insensitive `group by` and other string comparison operations

### ISSUE THREAD

**Issue by smudge202** [datalust/seq-tickets#687]
Group by clauses are case sensitive (which is fine) but I don't believe there is a way to make this case-insensitive and there are no functions (that I can find) to change the case of a field (`ToLower(` for example).

**Comment by nblumhardt** [comment]
Thanks for the note, @smudge202. Definitely a bit of a gap here - `ToLower()` would be a nice quick workaround, but wouldn't necessarily be the most efficient option; we'll have a look at this and see what we can come up with 👍

**Comment by nblumhardt** [comment]
@smudge202 I'm re-titling this because it's not something we can address without reviewing other aspects of the query language for consistency (we don't want to rely on `ToLower()` for this because of the possible perf impact).

Along with `group by case-insensitive`, we think a round of language changes would be beneficial, and we can migrate existing data to minimize the impact of this:

 - [ ] Remove support for `$0123abcd` event types
 - [ ] Allow simple text searches equivalent to `"hello"`, or `@"Hello"`, in SQL expressions/`where` clauses
 - [ ] Consider deprecating C#-style operators `==`, `!=`, `!`
 - [ ] SQL strings in filters support embedded newlines, in queries they do not - we can probably accept embedded newlines everywhere without major issues (though we'll have to take care in syntax error reporting)
 - [ ] Review use of case sensitive vs insensitive function calls
 - [ ] Case-sensitive `like`, i.e. `like` and `likei`, if we follow the PostgreSQL model
 - [ ] 'not in' support

**Comment by smudge202** [comment]
This obviously broadens the scope of my original issue which I'm sure fits with your master plan. Be interested to see which operations will be made case-insensitive and what syntax changes entail (not familiar with the Postgre model myself but won't take long to read up if that's what's being applied).

If you're deprecating C# style operators will we get SQL's `<>` operator for not, or still use the `Not()` syntax?

I've never got `in` to work before (resorting to `or/or/or...`). One of those topics really difficult to google for (google `seq in`) and I couldn't see the syntax in the docs. Anyway, hopefully spot correct syntax usage in release notes or something if you're doing `not in`!

**Comment by nblumhardt** [comment]
Thanks for the follow-up.

`X <> 'foo'` is already supported today, and `not X` also works for Boolean expressions.

`in` syntax is:

```sql
@Level in ['Error', 'Fatal']
```

Within Seq's data model this is a test against an array `[]` and JSON notation is used for those, unlike SQL's `in`, which uses the `()`. Googling is, sadly, almost impossible. It's recently been added [to the docs](https://docs.datalust.co/docs/query-syntax#section--in-expressions) but it's easy to miss, in there.

HTH!

**Comment by smudge202** [comment]
Derp. Square brackets, of course! Not sure if you heard that facepalm all the way over in Australia, but it was substantial.

Thanks. Can be a little tricky guessing whether to expect SQL or JSON syntax at times. I should remember to try both when in doubt.

**Comment by c0shea** [comment]
I actually prefer the C# style operators like `!=` to `<>`, since it works in SQL Server, too (yes, I know it's not ANSI standard SQL). Especially considering in uses square brackets instead of parenthesis, it already shows a deviation from SQL dialects.

**Comment by nblumhardt** [comment]
Thanks for the feedback, @c0shea 👍

**Comment by nblumhardt** [comment]
Thinking some more about how case-insensitive variants might be expressed (since there are a lot of different operations involving equality comparisons), we might be better off following the syntactic mechanism used for `escape` in `like` queries, which the parser will already need to accommodate.

E.g. if we chose `ci` as the modifier, we could semi-naturally express case-insensitive versions of various things:

```sql
x like 'foo%'
x like 'foo%' ci

group by a, b, c
group by a, b ci, c

a in ['b', 'c']
a in ['b', 'c'] ci

select distinct(a)
select distinct(a) ci

-- taking it a step further
a = 'b' and c = d
a = 'b' ci and c = d

-- an alternative to `@"some text"` that doesn't
-- clash as badly with SQL's use of `@` for identifier
-- prefixes?
"some text"
"some text" ci
```

It's a little early in the morning to consider what this would mean for more complex expressions, e.g. `(a < b or c = d) ci` - and what oddities that might introduce into the grammar... 🤔

**Comment by nblumhardt** [comment]
The examples

```sql
group by x ci
group by x and a = 'b' ci
```

reveal the tricky case of deciding where to apply the `ci` - in the latter case, the group key will be a Boolean and `ci` should be considered part of the `=` expression; in the former, regardless of `x`'s type at runtime, `ci` should be considered part of `group by`.

It's easy enough to deal with this, but it does introduce context sensitivity into the grammar, which deserves some careful consideration. No clean keyword-like alternatives seem any better - will keep on down this path and see what emerges :-)

**Comment by nblumhardt** [comment]
We've resolved the case above by treating `ci` as part of the operator, rather than "turning on case insensitivity within the scope of an expression", so:

```
group by x ci
group by x and a = 'b' ci
```

will both give the expected parses/behavior - the first is a case-insensitive grouping, the second is a case-insensitive comparison of `a` and `'b'`.

Making good progress on this, with the case insensitive modifier nearly ready to preview. Time to review the rest of this thread and tie it all together!

**Comment by nblumhardt** [comment]
Scoping this back down to just the `ci` modifier - I've written a quick RFC and posted it to:

https://github.com/datalust/seq-tickets/pull/919

Quite a few details in there, hopefully answers some questions (even if it poses a few more 😅 ) - would appreciate any feedback you can add!

**Comment by nblumhardt** [comment]
This is now implemented through the universal `ci` modifier.

A quick summary of where we landed: `ci` will turn practically _any_ string operation in Seq into a case-insensitive one:

```sql
x like 'foo%' ci

group by a, b ci, c

order by a ci desc

a in ['b', 'c'] ci

select distinct(a) ci

a = 'b' ci and c = d

IndexOf(a, 'b') ci
```

We ended up without too many syntactic edge cases; the main one is that `ci` binds tighter to function calls/other expressions than to grouping/ordering constructs, so to group by a substring in a case-insensitive fashion we have:

```sql
group by (Substring(a, 0, 1)) ci
```

With this feature implemented, `like` becomes case-sensitive by default; this is a breaking change, but existing signals/dashboards etc. will be updated automatically upon upgrade.

We left text fragment searches alone - `"some text"` is still a CI search; `@"some text"` remains but is deprecated - we think the `@Message like '%some text%'` alternative is less of a syntactic dead-end and will show this in the docs.

For the sake of interest, a heap of other good changes went through with this. We broke our dependency on Sprache (filters, queries, and other expressions now all use the same Superpower-based tokenizer and parser), we fixed various bugs, and filled gaps. The ones not directly related to case sensitivity, I'll raise additional tickets for.

Thanks @smudge202 and @c0shea for your help!
